### PR TITLE
[UNR-2945] Add gym to test asynchronous asset loading in FastArraySerialization-replicated arrays

### PR DESCRIPTION
**Pull request by ImprobableNic**
A gym to test what happens if references to assets that will be asynchronously loaded are serialized through a FastArraySerialize array.
What we want to avoid is a callback (PostReplicatedAdd) being issued on an item with unresolved reference.
The replicated cube on the center will inherit from ANetSerializeTestActor, which contains a FastArraySerialize array and methods to inspect it.
The material setter blueprint will be set on the map server-side with an array of soft references to material assets. It will set material references on the replicated cube and will display client side if it received a callback with null references.

Gym : 
image

Material setter blueprint : 

Client-side check
image

Server-side setup
image
image

FASAsyncTestActor blueprint

image

### CODE REVIEW

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
+#pragma once
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.h:1]: Spacing in this file seems off

**Comment by m-samiec** [comment]
> The material setter blueprint will be set on the map server-side with an array of soft references to material assets. It will set material references on the replicated cube and will display client side if it received a callback with null references.

Does the client ever load these references?

**Comment by m-samiec** [comment]
When is the `Client-side check` triggered?

**Comment by m-samiec** [comment]
I'm confused by the `divide by 2s` everywhere

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#include "DeltaSerializeStruct.h"
+#include "Net/UnrealNetwork.h"
+
+void FTestStructItem::PostReplicatedChange(const struct FTestStructContainer& InArraySerializer)
+{
+  int32 ElementIdx = this - InArraySerializer.Items.GetData();
+  if (Material)
+  {
+    int32 ChangeNum = InArraySerializer.Owner->ElementsHavingChanged.FindOrAdd(ElementIdx, 0);
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:9]: Doesn't this need to return a reference if you want to modify it afterwards?

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
+#include "DeltaSerializeStruct.h"
+#include "Net/UnrealNetwork.h"
+
+void FTestStructItem::PostReplicatedChange(const struct FTestStructContainer& InArraySerializer)
+{
+  int32 ElementIdx = this - InArraySerializer.Items.GetData();
+  if (Material)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:7]:
```suggestion
  if (Material != nullptr)
```

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#include "DeltaSerializeStruct.h"
+#include "Net/UnrealNetwork.h"
+
+void FTestStructItem::PostReplicatedChange(const struct FTestStructContainer& InArraySerializer)
```
Comment (m-samiec) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:4]: There are callbacks available for individual elements within a fas, couldn't that be used here instead?

Comment (m-samiec) [reply Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:4]: Which include the element idx, so you don't need to do the pointer magic

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp (hunk 3 above)
Comment (m-samiec) [reply Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:7]: Same for other Material checks around

**Comment by ImprobableNic** [comment]
@m-samiec 
Async loading is manually enabled in the level blueprint through a helper function.
The client side check is triggered on BeginPlay (the first three graph are wired on begin play, with a switch on auth).
The divide by two is to have two batches of materials, one to add elements, one to change elements.

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp (hunk 2 above)
Comment (ImprobableNic) [reply Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp]: That was some old prototype, removed now.

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+}
+
+void FTestStructContainer::PostReplicatedAdd(const TArrayView<int32>& AddedIndices, int32 FinalSize)
+{
+  Owner->NumInvalidItems = 0;
+
+  for (const auto& CurItem : Items)
+  {
+    if (CurItem.Material == nullptr)
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:20]: Nit - there's quite a few place that use 2 spaces instead of a tab.

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    if (Item.Material == nullptr)
+    {
+      ++NumInvalidItems_OnRep;
+    }
+  }
+
+  ReplicationHappened();
+}
+
+void ANetSerializeTestActor::GetLifetimeReplicatedProps(class TArray<class FLifetimeProperty> & OutLifetimeProps) const
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:49]: Other nit - prefer forward declarations a the top and use `Type&` rather than `Type &` .

**Comment by samiwh** [comment]
Can we add some instructions to this with some preamble about what to expect

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.h [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	UFUNCTION(BlueprintCallable)
+	void SetMaterial(int32 Index, UMaterial* InMaterial) 
+	{ 
+		TestContainer.Items[Index].Material = InMaterial; 
+		TestContainer.MarkItemDirty(TestContainer.Items[Index]);
+	}
+
+protected:
+
+	void GetLifetimeReplicatedProps(class TArray<FLifetimeProperty>&) const;
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.h:94]:
```suggestion
	void GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const;
```

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+{
+
+}
+
+void FTestStructContainer::PostReplicatedAdd(const TArrayView<int32>& AddedIndices, int32 FinalSize)
+{
+  Owner->NumInvalidItems = 0;
+
+  for (const auto& CurItem : Items)
+  {
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:19]: Nit using spaces instead of tabs in this file.

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    if (Item.Material == nullptr)
+    {
+      ++NumInvalidItems_OnRep;
+    }
+  }
+
+  ReplicationHappened();
+}
+
+void ANetSerializeTestActor::GetLifetimeReplicatedProps(class TArray<FLifetimeProperty> & OutLifetimeProps) const
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:49]:
```suggestion
void ANetSerializeTestActor::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
```

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.h [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+	bool NetDeltaSerialize(FNetDeltaSerializeInfo& DeltaParms);
+
+	void PreReplicatedRemove(const TArrayView<int32>& RemovedIndices, int32 FinalSize);
+	void PostReplicatedAdd(const TArrayView<int32>& AddedIndices, int32 FinalSize);
+	void PostReplicatedChange(const TArrayView<int32>& ChangedIndices, int32 FinalSize);
+};
+
+template<>
+struct TStructOpsTypeTraits< FTestStructContainer > : public TStructOpsTypeTraitsBase2< FTestStructContainer >
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.h:55]:
```suggestion
struct TStructOpsTypeTraits<FTestStructContainer> : public TStructOpsTypeTraitsBase2<FTestStructContainer>
```

File: USER_MANUAL.md [hunk 11]
```diff
@@ -218,5 +218,19 @@ Actors used in gyms are in `Contetnt\Actors`: add any new Actors to this directo
 * Pressing "enter" will print out information on the client regarding the owners of each cube. It will also send a server RPC that will print out if it was successfully processed on the server. 
 * Note that initially no cube is possessed so all server RPCs sent on the cubes will fail. After a cube is possessed, the server RPC called on a cube that does not have a player controller owner will not be processed.
 * Ensure load balancing is turned off.
+
+#### FASAsyncGym
+* Checks an edge case of the GDK handling of FastSerialized Arrays.
+* Native Unreal prevents FAS(FastArraySerialization) callbacks being called on null pointers to loadable assets.
+* This test creates a situation where an asset will be loaded on the backgrounds and pointers to it being replicated early.
```
Comment (samiwh) [inline USER_MANUAL.md:225]:
```suggestion
* This test creates a situation where pointers to an asset will be replicated before the asset has been loaded on the client.
```

File: USER_MANUAL.md [hunk 12]
```diff
@@ -218,5 +218,19 @@ Actors used in gyms are in `Contetnt\Actors`: add any new Actors to this directo
 * Pressing "enter" will print out information on the client regarding the owners of each cube. It will also send a server RPC that will print out if it was successfully processed on the server. 
 * Note that initially no cube is possessed so all server RPCs sent on the cubes will fail. After a cube is possessed, the server RPC called on a cube that does not have a player controller owner will not be processed.
 * Ensure load balancing is turned off.
+
+#### FASAsyncGym
+* Checks an edge case of the GDK handling of FastSerialized Arrays.
+* Native Unreal prevents FAS(FastArraySerialization) callbacks being called on null pointers to loadable assets.
```
Comment (samiwh) [inline USER_MANUAL.md:224]:
```suggestion
* Native Unreal prevents async asset loading causing null pointers in FAS (FastArraySerialization) callbacks.
```

File: USER_MANUAL.md [hunk 13]
```diff
@@ -218,5 +218,19 @@ Actors used in gyms are in `Contetnt\Actors`: add any new Actors to this directo
...
+* This test creates a situation where an asset will be loaded on the backgrounds and pointers to it being replicated early.
+* When async loading completes, the pointers should get valued, and FAS callbacks be called with valid pointers.
```
Comment (samiwh) [inline USER_MANUAL.md:226]:
```suggestion
* When async loading completes the FAS callbacks will be called with valid pointers.
```

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+#include "Net/UnrealNetwork.h"
+
+void FTestStructContainer::PostReplicatedChange(const TArrayView<int32>& ChangedIndices, int32 FinalSize)
+{
+	PostReplicatedAdd(ChangedIndices, FinalSize);
+}
+
+void FTestStructContainer::PreReplicatedRemove(const TArrayView<int32>& RemovedIndices, int32 FinalSize)
+{
+
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:11]:
```suggestion
```

File: Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+    if (CurItem.Material == nullptr)
+    {
+      ++Owner->NumInvalidItems;
+    }
+  }
+
+  Owner->ReplicationHappened();
+}
+
+bool FTestStructContainer::NetDeltaSerialize(FNetDeltaSerializeInfo & DeltaParms)
```
Comment (samiwh) [inline Game/Source/GDKTestGyms/DeltaSerializeStruct.cpp:29]:
```suggestion
bool FTestStructContainer::NetDeltaSerialize(FNetDeltaSerializeInfo& DeltaParms)
```
